design/spark-co-exist-with-object-storages.md

```diff
@@ -0,0 +1,195 @@
+
+##### Multiple mapping configurations   
+
+Each `toFsScheme` can have any number of mapping configurations. E.g., below are two mapping configurations for `toFsScheme=lakefs`.
+```properties
+routerfs.mapping.1.lakefs.replace='^s3a://bucket/prefix'
+routerfs.mapping.1.lakefs.with='lakefs://example-repo/dev/prefix'
+routerfs.mapping.2.lakefs.replace='^s3a://another-bucket'
+routerfs.mapping.2.lakefs.with='lakefs://another-repo/main'
+```
+Mapping configurations applied in order, therefore in case of a conflict in mapping configurations the prior configuration 
+applies. 
+
+##### Multiple `toFsScheme`s (and multiple mapping configuration groups)
+
+With `RouterFileSystem`, Spark users can define any number of `toFsScheme`s. Each forms its own mapping configuration group,
+and allows applying different set of Spark/Hadoop configurations. e.g. credentials, s3 endpoint, etc. Users would typically
+define new `toFsScheme` while trying to migrate a collection from one storage space to another without changing their Spark code.
+
+The example below demonstrates how routerFS mapping configurations and some Hadoop configurations will look like for
+Spark application that accesses s3, MinIO, and lakeFS, but is using s3a as its sole URI scheme. 
+```properties
+# Mapping configurations
+routerfs.mapping.1.lakefs.replace='^s3a://bucket/prefix'
+routerfs.mapping.1.lakefs.with='lakefs://example-repo/dev/prefix'
+routerfs.mapping.1.s3a1.replace='^s3a://bucket'
+routerfs.mapping.1.s3a1.with='s3a1://another-bucket'
+routerfs.mapping.1.minio.replace='^s3a://minio-bucket'
+routerfs.mapping.1.minio.with='minio://another-minio-bucket'
+
+# File System configurations
+fs.s3a.impl=RouterFileSystem
+fs.lakefs.impl=S3AFileSystem
+fs.s3a1.impl=S3AFileSystem
+fs.minio.impl=S3AFileSystem
+```
+
+##### Default mapping configuration 
+
+`RouterFileSystem` requires a default mapping configuration in case that none of the mappings matches a URI. 
+For example, the following default configuration states that in case or a URI with `scheme=s3a` that didn't match the
+`^s3a://bucket/prefix` pattern, `RouterFileSystem` uses the default file system that's configured to be `S3AFileSystem`.  
+
+```properties
+routerfs.mapping.1.lakefs.replace='^s3a://bucket/prefix'
+routerfs.mapping.1.lakefs.with='lakefs://example-repo/dev/prefix'
+
+# Default mapping, should be applied last 
+routerfs.mapping.default.replace='^s3a://'
+routerfs.mapping.default.with='default://'
+
+# File System configurations
+fs.s3a.impl=RouterFileSystem
+fs.lakefs.impl=S3AFileSystem
+fs.default.impl=S3AFileSystem
+```
+
+This configuration is required because otherwise `RouterFileSystem` will get stuck in an infinite loop, by calling itself 
+as the file system for `scheme=s3a` in the example. 
+
+**Note** Hadoop configurations include a `fs.default.impl` and we will need to test that this configuration is working as expected.
```

Comment:
From your examples I understand that the user can choose any scheme name for the default - so in the example I would use `s3a-wrapped`, `s3a1` or something less general than "default".

Reply:
how about using `s3a-actual`?

Reply:
I renamed to s3a-default and removed the note, you are right thanks for the suggestion.